WordPress/src/main/java/org/wordpress/android/ui/posts/PageAndPostListPreviewStateHelper.kt

```diff
@@ -0,0 +1,83 @@
+import org.wordpress.android.fluxc.model.PostModel
+import org.wordpress.android.ui.posts.RemotePreviewLogicHelper.RemotePreviewType
+import org.wordpress.android.ui.uploads.UploadService
+import org.wordpress.android.util.AppLog
+
+fun managePreviewStateTransitions(
+    newState: PostListRemotePreviewState,
+    prevState: PostListRemotePreviewState?,
+    postInfo: PostInfoType,
+    handleRemotePreview: (localPostId: Int, RemotePreviewType) -> Unit
+) = when (newState) {
+    PostListRemotePreviewState.PREVIEWING -> {
+        prevState?.let {
+            if (it == PostListRemotePreviewState.UPLOADING_FOR_PREVIEW ||
+                    it == PostListRemotePreviewState.REMOTE_AUTO_SAVING_FOR_PREVIEW) {
+                (postInfo as? PostInfoType.PostInfo)?.let { info ->
+                    handleRemotePreview(info.post.id, getRemotePreviewType(prevState))
+                }
+            }
+        }
+    }
+    PostListRemotePreviewState.NONE,
+    PostListRemotePreviewState.UPLOADING_FOR_PREVIEW,
+    PostListRemotePreviewState.REMOTE_AUTO_SAVING_FOR_PREVIEW,
+    PostListRemotePreviewState.REMOTE_AUTO_SAVE_PREVIEW_ERROR -> {
+        // nothing to do
+    }
+}
+
+fun getUploadStrategyFunctions(
+    activity: FragmentActivity,
+    action: PostListAction.PreviewPost
+) = object : RemotePreviewLogicHelper.RemotePreviewHelperFunctions {
+    override fun notifyUploadInProgress(post: PostModel): Boolean {
+        return if (UploadService.hasInProgressMediaUploadsForPost(post)) {
+            action.showToast.invoke(action.messageMediaUploading)
+            true
+        } else {
+            false
+        }
+    }
+
+    override fun startUploading(isRemoteAutoSave: Boolean, post: PostModel) {
+        if (isRemoteAutoSave) {
+            action.triggerPreviewStateUpdate.invoke(
+                    PostListRemotePreviewState.REMOTE_AUTO_SAVING_FOR_PREVIEW,
+                    PostInfoType.PostNoInfo
+            )
+            if (!UploadService.isPostUploadingOrQueued(post)) {
+                UploadService.uploadPost(activity, post, true)
+            } else {
+                AppLog.d(
+                        AppLog.T.POSTS,
+                        "Remote auto save for preview not possible: post already uploading or queued"
+                )
+            }
+        } else {
+            action.triggerPreviewStateUpdate.invoke(
+                    PostListRemotePreviewState.UPLOADING_FOR_PREVIEW,
+                    PostInfoType.PostNoInfo
+            )
+            if (!UploadService.isPostUploadingOrQueued(post)) {
+                UploadService.uploadPost(activity, post)
+            } else {
+                AppLog.d(
+                        AppLog.T.POSTS,
+                        "Upload for preview not possible: post already uploading or queued"
+                )
+            }
+        }
+    }
+}
+
+private fun getRemotePreviewType(prevState: PostListRemotePreviewState): RemotePreviewType {
```

Comment:
This private function seems a bit misleading. It returns `RemotePreviewType.REMOTE_PREVIEW_WITH_REMOTE_AUTO_SAVE` for `PostListRemotePreviewState.NONE` and `PostListRemotePreviewState.PREVIEWING`. I understand that it's called from a spot where we know it'll be either `PostListRemotePreviewState.UPLOADING_FOR_PREVIEW` or `PostListRemotePreviewState.REMOTE_AUTO_SAVING_FOR_PREVIEW`, but imho we should try to prevent the client from making a mistake. I'd consider throwing an exception if the method is called with an unexpected argument.

Reply:
Fixed in ad64d06d3915d0720af6a38fd6456e74ed0040af